kep/206-resource-limit-aware-scoring/README.md

```diff
@@ -0,0 +1,97 @@
+
+To reduce the extent of resource over-subscription and associated resource contention risks, we propose a scoring plugin that considers a node's resource limit in scheduling.
+
+## Goals
+Develop and add a scoring plugin to mitigate resource over-subscription issue caused by burstable pods through "spreading" or "balancing" pod's resource limits across nodes. 
+
+## Non-Goals
+Replace existing resource allocated plugin. 
+
+## Use Cases
+Consider two nodes with the same resource allocatable (8 CPU cores) and each has two pods running.
+- node1: pod1(request: 2, limit: 6), pod2(request: 2, limit: 4); total request/allocatable: 4/8, total limit/allocatable: 10/8
+- node2: pod3(request: 3, limit: 3), pod4(request: 2, limit: 2); total request/allocatable: 5/8, total limit/allocatable: 5/8
+
+**Default scheduler**
+
+To schedule a new pod pod5 (request: 1, limit: 4), the default scheduler with `LeastAllocated` will choose node1 over node2 as node1 has has a smaller request/allocatable ratio (4/8 vs. 5/8). The resulting node resource status will be
+- node1: request/allocatable = 5/8, limit/allocatable = 14/8  
+- node2: request/allocatable = 5/8, limit/allocatable = 5/8
+
+node1's resource limit is oversubscribed with a limit to allocatable ratio 14/8.  It will have a high risk of resource contention.
+
+**Proposed scheduler**
+
+If we consider node1's resource limit is already oversubscribed and hence place pod5 on node 2 instead, the result will be 
+- node1: total request/allocatable = 4/8, total limit/allocatable = 10/8  
+- node2: total request/allocatable = 6/8, total limit/allocatable = 9/8
+
+node1’s total resource limit remains the same and node2 has a total resource limit of 9. node2 will have a lower risk of resource contention with pod5 running on it. 
+
+
+## Terms
+
+## Proposal
+
+A custom scheduler scoring plugin scores a node by considering existing pods’ resource limits to the node's allocatable ratio on the node. It attempts to “evenly” spread or distribute pod resource limits across nodes in a cluster. 
+
+## Design Details
+
+When choosing a node to place a pod, the custom scoring plugin prefers the node that has the smallest total resource limit ratio on a node. The scoring function favors a node with the smallest over-subscription ratio.
+
+   `node limit ratio = sum of existing pods' limits on node / node’s allocatable resource`
+
+1. Calculate each node’s raw score. A node's raw score will be negative for an over-subscription node.
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
```

Comment:
Do we need to think about best effort pods here？

Reply:
> Do we need to think about best effort pods here？

Very good question. There are two options.

1. Skip best effort pods.

2. Use allocatable as best effort pods's limits. 

BTW, I suspect best effort pods are not allowed in most production environments.

Reply:
The default can be the node's allocatable.